File: day7/soln.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type bags struct {
	count int
	colour string
}

func createGraphmap(ipt []string) map[string][]bags {
	graphmap := make(map[string][]bags)

	for _, i := range ipt {
		var children []bags
		tokens := strings.Split(i, " ")
		parent := strings.Join(tokens[:2], " ")

		if !strings.Contains(i, "no other bags") {
			for i := 4; i < len(tokens); i += 4 {
				num, _ := strconv.Atoi(tokens[i])
				children = append(children, bags{
					num,
					strings.Join(tokens[i+1:i+3], " "),
				})
			}
		}

		graphmap[parent] = children
	}

	return graphmap
}

func part1(ipt []string, target string) (total int) {
	graphmap := createGraphmap(ipt)

	var searchGraph func(string) int
	searchGraph = func(root string) int {
		for _, c := range graphmap[root] {
			if strings.Contains(c.colour, target) || searchGraph(c.colour) > 0 {
				return 1
			}
		}

		return 0
	}

	for k := range graphmap {
		total += searchGraph(k)
	}

	return
}

func part2(ipt []string, target string) int {
	graphmap := createGraphmap(ipt)

	var searchGraph func(string) int
	searchGraph = func(root string) (total int) {
		for _, c := range graphmap[root] {
			total += c.count + c.count * searchGraph(c.colour)
		}

		return
	}

	return searchGraph(target)
}

func main() {
	var ipt []string

	file, _ := os.Open("input.txt")
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		ipt = append(ipt, scanner.Text())
	}

	fmt.Println(part1(ipt, "shiny gold"))
	fmt.Println(part2(ipt, "shiny gold"))
}

File: day7/soln.py
def create_graphmap(ipt):
    graphmap = {}

    for i in ipt:
        tokens = i.split(' ')
        parent = ' '.join(tokens[:2])
        children = []

        if not 'no other bags' in i:
            for i in range(4, len(tokens), 4):
                children.append((int(tokens[i]), ' '.join(tokens[i+1:i+3])))

        graphmap[parent] = children

    return graphmap


def part1(ipt, target):
    graphmap = create_graphmap(ipt)

    def search_graph(root):
        for c in graphmap[root]:
            if target in c[1] or search_graph(c[1]):
                return 1

        return 0

    return sum([search_graph(k) for k, _ in graphmap.items()])


def part2(ipt, target):
    graphmap = create_graphmap(ipt)

    def search_graph(root):
        total = 0
        for c in graphmap[root]:
            total += c[0] + c[0] * search_graph(c[1])

        return total

    return search_graph(target)


if __name__ == '__main__':
    with open('input.txt') as fp:
        ipt = [l for l in fp]
        print(part1(ipt, 'shiny gold'))
        print(part2(ipt, 'shiny gold'))
